flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/KinesisStreamsSourceITCase.java

```diff
@@ -0,0 +1,218 @@
+import software.amazon.awssdk.services.kinesis.model.StreamStatus;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.apache.flink.connector.aws.config.AWSConfigConstants.AWS_ACCESS_KEY_ID;
+import static org.apache.flink.connector.aws.config.AWSConfigConstants.AWS_ENDPOINT;
+import static org.apache.flink.connector.aws.config.AWSConfigConstants.AWS_REGION;
+import static org.apache.flink.connector.aws.config.AWSConfigConstants.AWS_SECRET_ACCESS_KEY;
+import static org.apache.flink.connector.aws.config.AWSConfigConstants.HTTP_PROTOCOL_VERSION;
+import static org.apache.flink.connector.aws.config.AWSConfigConstants.TRUST_ALL_CERTIFICATES;
+
+/** IT cases for using {@code KinesisStreamsSource} using a localstack container. */
+@Testcontainers
+@ExtendWith(MiniClusterExtension.class)
+public class KinesisStreamsSourceITCase {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KinesisStreamsSourceITCase.class);
+    private static final String LOCALSTACK_DOCKER_IMAGE_VERSION = "localstack/localstack:3.7.2";
+
+    @Container
+    private static final LocalstackContainer MOCK_KINESIS_CONTAINER =
+            new LocalstackContainer(DockerImageName.parse(LOCALSTACK_DOCKER_IMAGE_VERSION));
+
+    private StreamExecutionEnvironment env;
+    private SdkHttpClient httpClient;
+    private KinesisClient kinesisClient;
+
+    @BeforeEach
+    void setUp() {
+        System.setProperty(SdkSystemSetting.CBOR_ENABLED.property(), "false");
+
+        env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setParallelism(1);
+
+        httpClient = AWSServicesTestUtils.createHttpClient();
+        kinesisClient =
+                AWSServicesTestUtils.createAwsSyncClient(
+                        MOCK_KINESIS_CONTAINER.getEndpoint(), httpClient, KinesisClient.builder());
+    }
+
+    @AfterEach
+    void teardown() {
+        System.clearProperty(SdkSystemSetting.CBOR_ENABLED.property());
+        AWSGeneralUtil.closeResources(httpClient, kinesisClient);
+    }
+
+    @Test
+    void nonExistentStreamShouldResultInFailure() {
+        Assertions.assertThatExceptionOfType(RuntimeException.class)
+                .isThrownBy(
+                        () ->
+                                new Scenario()
+                                        .localstackStreamName("stream-exists")
+                                        .withSourceConnectionStreamArn(
+                                                "arn:aws:kinesis:ap-southeast-1:000000000000:stream/stream-not-exists")
+                                        .runScenario())
+                .havingCause()
+                .havingCause()
+                .havingCause()
+                .havingCause()
+                .havingCause()
+                .havingCause()
+                .havingCause()
```

Comment:
Could we check this recursively instead?

Reply:
Good point, I've simplified the assertion using `.withStackTraceContaining`. We no longer assert the exact position of the cause in the stack trace which is more robust.